hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/contracts/execution/traceability/HederaTracer.java

```diff
@@ -60,29 +78,68 @@ public class HederaTracer implements HederaOperationTracer {
     private static final int OP_CODE_CREATE2 = 0xF5;
     private static final int OP_CODE_STATICCALL = 0xFA;
 
-    public HederaTracer(final boolean areActionSidecarsEnabled) {
+    public enum EmitActionSidecars {
+        DISABLED,
+        ENABLED
+    }
+
+    public enum ValidateActionSidecars {
+        DISABLED,
+        ENABLED
+    }
+
+    public HederaTracer(
+            final EmitActionSidecars emitActionSidecars, final ValidateActionSidecars validateActionSidecars) {
         this.currentActionsStack = new ArrayDeque<>();
         this.allActions = new ArrayList<>();
-        this.areActionSidecarsEnabled = areActionSidecarsEnabled;
+        this.invalidActions = new ArrayList<>();
+        this.doEmitActionSidecars = emitActionSidecars;
+        this.doValidateActionSidecars = validateActionSidecars;
     }
 
     @Override
     public void init(final MessageFrame initialFrame) {
-        if (areActionSidecarsEnabled) {
+        if (areActionSidecarsEnabled()) {
             trackTopLevelActionFor(initialFrame);
         }
     }
 
+    @Override
+    public void finalizeOperation(final MessageFrame initialFrame) {
+        if (areActionSidecarsEnabled() && isActionSidecarValidationEnabled()) {
+            // Two possible error conditions are that invalid actions (mainly `oneof` fields that
+            // don't have _exactly_ one alternative set) are added to the set of actions (which can
+            // cause problems when ingesting sidecar records elsewhere) and that the current actions
+            // stack is out-of-sync (due to some error elsewhere).  (The latter condition has not yet
+            // been observed outside deliberate fault injection and probably doesn't happen in a
+            // running system.)
+            if (!currentActionsStack.isEmpty() || !invalidActions.isEmpty()) {
+                log.atLevel(logLevel)
```

Comment:
Log level is debug by default, this is too low.  Should be warn.  If we are executing this code either (a) we have the actions code wrong and need to fix it or (b) something else has gone horribly wrong in the EVM for the stack to be out of sync, and warrants disabling the feature at runtime.  Defense in depth is great, but we should be aiming to get the code executing this block never because we have the action code right.  so the code can stay, but we need to acknowledge this is not an ignorable error.